bsip-0047.md

```diff
@@ -0,0 +1,182 @@
+* If any extension is present, it must contain the ID of an existing account.
+* If any extension is present, `account_options.voting_account == GRAPHENE_PROXY_PER_CATEGORY_ACCOUNT`
+
+
+### `account_object`(Existing)
+
+**Note:** this is an implementation detail and meant to be a hint for later development. The actual implementation may differ.
+
+The account object shall receive these new fields:
+
+* `flat_set<vote_id_type> committee_votes`
+* `flat_set<vote_id_type> witness_votes`
+* `flat_set<vote_id_type> worker_votes`
+
+`account_create_operation` and `account_update_operation` shall be modified to create these sets by filtering the `account_options.votes` field appropriately. `account_update_votes_operation` shall also create these sets by filtering the `votes_to_add` and `votes_to_remove` fields appropriately.
+
+
+### `account_update_votes_operation` (New)
+
+The following is an example implementation.
+
+```
+struct account_update_votes_operation : public base_operation
+{
+	asset fee;
+
+	/// The account to update
+	account_id_type account;
+
+	/// New account options
+	flat_set<vote_id_type>    votes_to_add;
+	flat_set<vote_id_type>    votes_to_remove;
+
+	// A new voting account to set
+	optional<account_id_type> committee_voting_account
+	optional<account_id_type> witness_voting_account
+	optional<account_id_type> worker_voting_account
+
+	// A new number of witness_votes to set
+	optional<uint16_t> num_witness;
+
+	// A new number of committee_votes to set
+	optional<uint16_t> num_committee;
+
+	// For future extensions (see account_update_operation)
+	extensions_type extensions;
+};
+```
+
+Changes to direct votes can be specified by means of a *list* of vote identifiers to be added or removed. This has multiple advantages:
+
+* Reduces the size of the operation compared to replacing the entire slate
+* Allows to easily compare votes before and after the operation
+
+**Validation checks:**
+
+* `account` must be a valid account ID, must exist, must authorize the operation, must pay fee
+* `committee_voting_account`, if present, must be a valid account ID, must exist
+* `witness_voting_account`, if present, must be a valid account ID, must exist
+* `worker_voting_account`, if present, must be a valid account ID, must exist
+* Validate the existence of each vote to add and remove
+* `num_witness`, if present, must be greater than or equal to  &ge; 0
+* `num_committee`, if present, must be greater than or equal to  &ge; 0
+
+**Evaluation:**
+
+* Update the account's `account_options.committee_votes`, `account_options.witness_votes`, and `account_options.worker_votes` by filtering the `votes_to_add` and `votes_to_remove` field appropriately
+* Update the `account_options.num_witness` if `witness_voting_account = GRAPHENE_PROXY_TO_SELF_ACCOUNT` and if `num_witness` is present
+* Update the `account_options.num_committee` if `committee_voting_account = GRAPHENE_PROXY_TO_SELF_ACCOUNT` and if `num_committee` is present
```

Comment:
If `votes_to_add` and `votes_to_remove` are accepted regardless of `GRAPHENE_PROXY_TO_SELF_ACCOUNT`, then `num_witness` and `num_committee` should be, too.

Reply:
I think you misunderstood me. That was not meant to be a new rule in the list, it was a comment hinting at an inconsistency in the existing list. Sorry for being inconsistent with my comment style. :-)

What I meant was: 
In step 1 you update the account's `*_votes` fields unconditionally. In steps 2 and 3 you update `num_witness` and `num_committee` only "if `witness_voting_account = GRAPHENE_PROXY_TO_SELF_ACCOUNT`.
You should either apply the condition to all steps or to none (IMO none is better).

Reply:
@pmconrad I did misunderstand so I appreciate the clarification.

I do think that we do need some form of the following checks:

* Update committee votes if and only if `account_options.witness_voting_account == GRAPHENE_PROXY_TO_SELF_ACCOUNT` **after** the merging of this operation
	* `committee_voting_account` may or may not be specified in the operation
* Update witness votes if and only if `account_options.witness_voting_account == GRAPHENE_PROXY_TO_SELF_ACCOUNT` **after** the merging of this operation
	* `witness_voting_account` may or may not be specified in the operation
* Update worker votes if and only if `account_options.worker_voting_account == GRAPHENE_PROXY_TO_SELF_ACCOUNT` **after** the merging of this operation
	* `worker_voting_account` may or may not be specified in the operation
* Update the `account_options.num_witness` if `num_witness` is present and if `account_options.witness_voting_account = GRAPHENE_PROXY_TO_SELF_ACCOUNT` **after** the merging of this operation
* Update the `account_options.num_committee` if `num_committee` is present and if `account_options.committee_voting_account = GRAPHENE_PROXY_TO_SELF_ACCOUNT`  **after** the merging of this operation

Take for example, 

* Update the `account_options.num_witness` if `num_witness` is present and if `account_options.witness_voting_account = GRAPHENE_PROXY_TO_SELF_ACCOUNT` **after** the effect of this operation

The trickier use cases are ones where one operation might update `num_witness` **or**  `witness_voting_account` **but not both** and then _merge_ with its counterpart that is missing in the operation but is already present in the account_options on the blockchain.  This type of check is essential but could be performed either in validation or evaluation depending on your recommendation.

Reply:
Ok, then you should add these constraints to the validation section. I. e. operations with these inconsistencies are not allowed into the blockchain.

Reply:
The updated validation specs attempt to cover the proper merging of the changes with the current values on the blockchain.

The updated and previous validation also permit an operation that has nothing of substance (an "empty" update).  This might be appropriate if the extensions field might be used.  Such an "empty" operation will have the practical effect of updating the account's last vote time.  We should add an extra validation if "empty" updates are to be prohibited.